### ISSUE TITLE
refactoring

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI-dev#25]
Some suggestions for making the code clearer, and perhaps for getting a speed up if something is running on the CPU that could be running on the GPU:

- [x] better avoid mentioning "cuda" in scvi.py (e.g. `torch.is_cuda_available`, `.cuda`).  Instead, call `vae.cuda()` right after instantiating `vae` if cuda is available. In scvi.py, use `new` and `torch.register_buffer` as necessary.
- [x] The code `eps = np.random.normal(0, 1, mu.shape)` runs on the CPU. Use torch commands instead.
- [x] make `Encoder` and `Decoder` separate classes that `VAE` instantiates. It's a nice way to break up the VAE class into smaller components. (The random sampling remains in the `VAE` class.)
- [x] Use pytorch's initializers rather the scipy's `truncnorm`. If possible, good to remove all numpy and scipy dependencies from `scvi.py`.
- [x] make the loss a method in the VAE class, rather that putting it in `train.py` . Use flags and/or  inheritance if we need multiple loss functions (e.g. zinb loss, nb loss).
- [x] need to call `vae.eval` somewhere for computing log likelihood. Also, good to print the training and testing log likelihood every 10 epochs or so during training.
- [x] Move `compute_log_likelihood` to `benchmark.py` from `train.py` -- likelihood isn't really about training.
- [x] Use the "csv" instead of "pandas", to avoid a heavyweight dependency. Remove pandas from `requirements.txt`.
- [x] make `train_test_split` a `@staticmethod` of `Dataset` (and remove `utils.py`).

**Comment by Edouard360** [comment]
Thanks for all these insights Jeffrey ! 
Branch edouard/refactoring should deal with the last 3, and we are working on the others with Maxime

**Comment by Edouard360** [comment]
NearestNeighbors from sklearn uses scipy so the build failed on my branch. Actually I removed it in last commit, but let's hope we need sklearn for the rest of the project 😄 Build now passes with 100% coverage.

**Comment by jeff-regier** [comment]
It's OK to use sklearn and scipy if we can't easily avoid using them. The suggestion about removing scipy and numpy dependencies above was just for the `scvi.py` file. That file specifies the neural network so good to use just PyTorch there whenever possible.

**Comment by jeff-regier** [comment]
I meant to mention this today:
http://pytorch.org/docs/master/nn.html#torch.nn.Module.eval

Usually log likelihood scores are better if you call it to turn off batch norm and dropout during testing.

**Comment by Edouard360** [comment]
I will change to  eval thanks !

Also I changed to register buffer:

https://github.com/YosefLab/scVI-dev/blob/d6bab08c981ac87b2d492b0d1604442dc90a709a/scvi/scvi.py#L28-L29

But I still have to change it as a variable for computation afterwards: is this a problem ?

https://github.com/YosefLab/scVI-dev/blob/d6bab08c981ac87b2d492b0d1604442dc90a709a/scvi/scvi.py#L65

https://github.com/YosefLab/scVI-dev/blob/99e5b799c632964ad03207b9158c62b43954caf3/scvi/log_likelihood.py#L7-L10

I forgot to remove torch.cuda here which I will do in next commit.

However, I don't know if I should keep it in benchmark.py also:

https://github.com/YosefLab/scVI-dev/blob/99e5b799c632964ad03207b9158c62b43954caf3/scvi/train.py#L4-L7

Because I only need it for kl_ponderation...

https://github.com/YosefLab/scVI-dev/blob/99e5b799c632964ad03207b9158c62b43954caf3/scvi/train.py#L28

**Comment by jeff-regier** [comment]
You can call `register_buffer` on a `Variable`, e.g.
```
        grid0_data = torch.from_numpy(np.mgrid[(-r):(r+1), (-r):(r+1)].transpose([2, 1, 0]))
        grid0 = Variable(grid0_data, requires_grad=False)
        self.register_buffer("grid0", grid0)
```

**Comment by jeff-regier** [comment]
I think you could just pass `kl_ponderation` as a python float to vae, without converting it to `dtype`.

Another issue is whether we want to keep any of this `kl_ponderation` stuff.

**Comment by Edouard360** [comment]
Ah yes I didn't know it could work without dtype. I will make the change for now.

**Comment by maxime1310** [comment]
The kl_ponderation term comes from Romain's code: he uses a warmup scheme where the ponderation of the KLD is a linear function of the number of epochs until it reaches one. I think perhaps this is because in some papers it's stated that  if you don't do this at the beginning of training you might get stucked in local minimas? But anyway it was a bit absurd to use a torch Variable and not a python float.

**Comment by Edouard360** [comment]
Normally, last commit should satisfy all requirements

**Comment by jeff-regier** [comment]
Thanks for making all the changes!
